community/kernel/src/main/java/org/neo4j/GraphDatabaseBuilder.java

```diff
@@ -0,0 +1,395 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.neo4j.graphdb.factory.GraphDatabaseSettings;
+import org.neo4j.helpers.ArrayUtil;
+import org.neo4j.helpers.Service;
+import org.neo4j.kernel.extension.KernelExtensionFactory;
+import org.neo4j.logging.LogProvider;
+import org.neo4j.logging.NullLogProvider;
+
+/**
+ * A builder for a GraphDatabase.
+ *
+ * @param <BUILDER> The concrete type of the builder
+ * @param <GRAPHDB> The concrete type of the GraphDatabase being built
+ */
+abstract class GraphDatabaseBuilder<BUILDER extends GraphDatabaseBuilder<BUILDER,GRAPHDB>, GRAPHDB extends GraphDatabase>
+{
+    private LogProvider logProvider;
+    protected final Map<String,String> params = new HashMap<>();
+    protected List<KernelExtensionFactory<?>> kernelExtensions = new ArrayList<>();
+
+    protected GraphDatabaseBuilder()
+    {
+        withLogProvider( NullLogProvider.getInstance() );
+        params.put( "ephemeral", "false" );
+        for ( KernelExtensionFactory factory : Service.load( KernelExtensionFactory.class ) )
+        {
+            kernelExtensions.add( factory );
+        }
+    }
+
+    protected abstract BUILDER self();
+
+    protected abstract GRAPHDB newInstance(
+            File storeDir,
+            LogProvider logProvider,
+            Map<String,String> params,
+            List<KernelExtensionFactory<?>> kernelExtensions );
+
+    /**
+     * Specify a {@link LogProvider} that will be used for logging within the graph database.
+     *
+     * @param logProvider a {@link LogProvider} to use for logging
+     * @return this builder
+     */
+    public BUILDER withLogProvider( LogProvider logProvider )
+    {
+        this.logProvider = logProvider;
+        return self();
+    }
+
+    /**
+     * Allow this graph database instance to upgrade the store if required.
+     *
+     * @return this builder
+     */
+    public BUILDER allowStoreUpgrade()
+    {
+        params.put( GraphDatabaseSettings.allow_store_upgrade.name(), "true" );
+        return self();
+    }
+
+    /**
+     * Open this graph database in read-only mode, allowing only read operations.
+     *
+     * @return this builder
+     */
+    public BUILDER readOnly()
+    {
+        params.put( GraphDatabaseSettings.read_only.name(), "true" );
+        return self();
+    }
+
+    /**
+     * Set the default cypher parser version.
+     *
+     * @param version The version number for the cypher parser version to use by default
+     * @return this builder
+     */
+    public BUILDER usingCypherParserVersion( String version )
```

Comment:
Still uses `using`. I'd like the consistency of `with`, better to be careful here since methods introduced here will be user facing. Also a couple of other methods below.